apps/prairielearn/src/migrations/20231217153015_lti13_course_instances__create.sql

```diff
@@ -0,0 +1,26 @@
+CREATE TABLE IF NOT EXISTS
+  lti13_course_instances (
+    --ags_lineitems text,
+    --ags_lineitem text,
+    context_id text NOT NULL,
+    context_label text,
+    context_title text,
+    course_instance_id BIGINT NOT NULL REFERENCES course_instances ON DELETE CASCADE ON UPDATE CASCADE,
+    created_at timestamptz NOT NULL DEFAULT current_timestamp,
+    deleted_at timestamptz,
+    deployment_id text NOT NULL,
+    id bigserial PRIMARY KEY,
+    lti13_instance_id BIGINT REFERENCES lti13_instances ON DELETE CASCADE ON UPDATE CASCADE
+    --nrps_context_memberships_url text,
+    --preferences jsonb,
+  );
+
+ALTER TABLE lti13_course_instances
+ADD CONSTRAINT lti13_course_instances_unique_idx UNIQUE NULLS NOT DISTINCT (
+  lti13_instance_id,
+  deployment_id,
+  context_id,
+  deleted_at
+);
```

Comment:
The trickiest part of this PR was figuring out how to do the unique constraints across the LTI standards, and soft deletes on this `lti13_course_instances` table.

https://www.imsglobal.org/spec/lti/v1p3#lti-domain-model seems like the mix of `iss`, `client_id`, `deployment_id`, `context_id`, and maybe even deeper things like `resource_link_id`. The `iss` and `client_id` are tied to our `lti13_instance_id`. The `NULLS NOT DISTINCT` feature was nice to make it work as expected with the soft delete table. I think this is all good, but I might need another check. (Or we learn and refactor as we get more real world data.)

I don't currently have anything in the UX that does anything with the soft deleted parts of the table. Other than the dormant data, I'm treating it like a hard delete table. I presume that's OK, although maybe we want this to be a hard delete table. I'm not sure how we make that distinction these days.

Reply:
In PT, we've definitely been erring hard towards hard deletes + audit logs instead of soft deletes. I'm inclined to do the same here unless you think there's any reason soft deletes could be useful. cc @mwest1066 for when he's back from traveling.

Reply:
From further discussion, we decided this could be hard deleted with audit logs. A recent commit changed this PR to do that.